### PR TITLE
Spring security

### LINKED ISSUE #10: Implement Spring Security
Implement:

- login functionality
- secure endpoints
- default admin creation

### PR DESCRIPTION
**Pull request by VladHolobyn**
Implemented with JWT
Closes #10

### CODE REVIEW

File: src/main/java/com/pnudev/librarysystem/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    @Bean
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        http
+                .csrf().disable()
+                .formLogin().disable()
+                .httpBasic().disable()
+                .sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .addFilterAfter(jwtTokenFilter, UsernamePasswordAuthenticationFilter.class)
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/config/SecurityConfig.java:38]: So basically each request gets verified through JWT filters?

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java [hunk 2]
```diff
@@ -27,33 +28,37 @@ public class CategoryController {
     private final CategoryService categoryService;
 
 
+    @Secured({"ADMIN", "CLIENT"})
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Better use `@PreAuthorize` as it allows you to work with SpEL

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java [hunk 3]
```diff
@@ -27,33 +28,37 @@ public class CategoryController {
...
-    ){
+    ) {
         return categoryService.searchCategoryByName(
                 name,
                 pageNumber.filter(p -> p > 0).orElse(0),
                 size.filter(s -> s > 0).orElse(5)
         );
     }
 
+    @Secured("ADMIN")
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Same here

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java [hunk 4]
```diff
@@ -27,33 +28,37 @@ public class CategoryController {
...
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
-    public CategoryDTO addCategory(@Valid @RequestBody CategoryDTO categoryDTO){
+    public CategoryDTO addCategory(@Valid @RequestBody CategoryDTO categoryDTO) {
         return categoryService.addCategory(categoryDTO);
     }
 
+    @Secured("ADMIN")
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Same here

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java [hunk 5]
```diff
@@ -27,33 +28,37 @@ public class CategoryController {
...
     @PutMapping
-    public CategoryDTO updateCategory(@Valid @RequestBody CategoryDTO categoryDTO){
+    public CategoryDTO updateCategory(@Valid @RequestBody CategoryDTO categoryDTO) {
         return categoryService.updateCategory(categoryDTO);
     }
 
+    @Secured("ADMIN")
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Same here

File: src/main/java/com/pnudev/librarysystem/mapper/UserMapper.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.pnudev.librarysystem.security.UserDetailsImpl;
+import com.pnudev.librarysystem.entity.User;
+import org.mapstruct.Mapper;
+
+
+@Mapper(componentModel = "spring")
+public interface UserMapper {
+
+    UserDetailsImpl toUserDetails(User user);
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/mapper/UserMapper.java:11]: pass interface, not implementation

File: pom.xml [hunk 7]
```diff
@@ -54,6 +54,20 @@
...
+            <artifactId>spring-boot-starter-security</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt</artifactId>
+            <version>0.9.1</version>
+        </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
+            <artifactId>jaxb-api</artifactId>
```
Comment (justme1135) [inline pom.xml:68]: Why do you need this lib?

File: src/main/resources/application.yaml [hunk 8]
```diff
@@ -9,3 +9,11 @@ spring:
 file:
   upload-dir: "./uploads"
   image-types: jpg, jpeg, png
+
+application:
+  security:
+    secret-key: ${SECRET_KEY}
+    token-expiration-time: 36000000
```
Comment (getmane) [inline src/main/resources/application.yaml]: break down long integers using _

File: src/main/java/com/pnudev/librarysystem/security/JwtTokenFilter.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+
+
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/security/JwtTokenFilter.java]: Please be consistent with your newlines (remove 1)

File: src/main/java/com/pnudev/librarysystem/security/UserDetailsImpl.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/security/UserDetailsImpl.java:42]: Why is a single implementation returning constant values? 
What are those supposed to represent?

File: src/main/java/com/pnudev/librarysystem/config/SecurityConfig.java (hunk 1 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/config/SecurityConfig.java:38]: Only those that need authentication: 
` if (authHeader == null || !authHeader.startsWith("Bearer "))  filterChain.doFilter(request, response);`
Or what do you mean?

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java (hunk 2 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Refactored

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java (hunk 3 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Refactored

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java (hunk 4 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Refactored

File: src/main/java/com/pnudev/librarysystem/controller/CategoryController.java (hunk 5 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/controller/CategoryController.java]: Refactored

File: src/main/java/com/pnudev/librarysystem/mapper/UserMapper.java (hunk 6 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/mapper/UserMapper.java:11]: Are you sure about this? So I need to implement a factory method?

File: pom.xml (hunk 7 above)
Comment (VladHolobyn) [reply pom.xml:68]: I had the same [problem ](https://stackoverflow.com/questions/55606519/getting-exception-java-lang-noclassdeffounderror-could-not-initialize-class-jav)

File: src/main/resources/application.yaml (hunk 8 above)
Comment (VladHolobyn) [reply src/main/resources/application.yaml]: Refactored

File: src/main/java/com/pnudev/librarysystem/security/JwtTokenFilter.java (hunk 9 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/security/JwtTokenFilter.java]: Refactored

**Review by getmane — APPROVED** [review]

**Pull request merged**